Add some Rust library building infrastructure

chrysn opened the pull request:
### Contribution description

Building on #16274 this adds a single module that can encapsulate arbitrarily many Rust-based modules.

It's ~~heavily WIP, and will be rebased~~, but the core idea is something that I already want to have discussable:

* For every RIOT build, there should be at most one Rust crate built and statically linked. That allows cargo to manage shared dependencies well, and avoid code like formatting being pulled in twice.
* The module introduced here is intended to serve as that module if there if crates should come in but no application Rust code is used.
* Conceptually, both the RIOT modules and the Rust crates stay separate:
  * Pure-Rust RIOT modules are pseudomodules on the RIOT side and pull in the single module.
  * The module checks the list of enabled modules, and pulls in the relevant crates.

Note that Rust-only modules have little options to be usable, as otherwise they won't be called into:
* They can export XFA sections (like the demo does).
* They can export implementations of otherwise weak symbols.
* They can do what other modules do, and add `extern int module_setup(void);` to wherever in the init system they need to be called initially. (Personally I'd discourage that in favor of XFA.)

### Testing procedure

It's not a full test, but it's a good one:

`$ make BOARD=microbit-v2 -C examples/saul all flash term`

### ~~TBD~~

~~It's all a bit rough so far...~~

but these items are still open:

* [ ] Become less hard-coded -- right now adding a module means changing Cargo.toml, src/lib.rs and the Makefiles
* [ ] Be available for application Rust code: Application Rust code should easily be able to take over this module's responsibilities.

### Issues/PRs references

This is based on #16274, ~~and will be rebased onto that. Only look at the commits not on that other PR, currently it's the last one only.~~

[edit: testing procedure]

chrysn commented:
Rebased to track #16274

chrysn commented:
Update: This is now all much simpler, thanks to the enhancements the regular Rust build process received throughout review process.

There is now a (currently short) list of pseudomodules that all pull in the `rust_lib_catchall` module. The only current one is `rust_module_examples_shell_commands`, the shell command demo from the module examples crate. It adds a `ps.rs` command, which is just `ps` but implemented in Rust (with different output, but good enough for a demo).

A test now uses this (based on the shell test), and completed successfully on native, particle-xenon and microbit-v2.

For a more practical test (like "try it out!"), take any example with a shell, add `USEMODULE += rust_module_examples_shell_commands`, and see the new command pop up in help.

In terms of the original TBD items, there's still some to be done, but not all:

* The list of crate dependencies that becomes the pseudomodules is probably best managed manually for the time being. Right now there's one module, and until there are not at least five, any time spent automating things there is wasted in not knowing what will come up.
* Making this work with Rust-in-application crates needs improvement. Right now it starts two separate cargo runs (one for each "module"), and that results in duplicate symbols. (But even if it were not, two-ships-in-the-night Rust modules would duplicate a lot of code). Fixing this will be twofold:
  * One mechanism will lock the new module out from actually building (while still gathering all the flags from the pseudomodules).
  * One mechanism will allow an application to swoop up all the functionality that would have been in the catchall module (or the module as a whole).

chrysn commented:
Disabling builds pending https://github.com/kaspar030/dwq/pull/3 -- only when that's through I see a chance to understand the build failures. (Addressing the module-and-application-at-the-same-time issues meanwhile).

chrysn commented:
Now extended to have Rust application work together with modules, and the process documented.

It's not overly pretty ("Add a dependency to this local crate"), but I rather have it this way and visible than automagic (where something in the build system adds the crate sneakily).

As for testing (the latest commit provides some but it's (ab)using an example), I'm considering a simple RIOT module that can be written in Rust and added -- maybe a software implementation of a SAUL LED that reports to stdout? (That'd actually be convenient to have in the SAUL example on native, given that Rust should be unproblematic on native, it's nothing that people can't live without, but it makes the saul example actually do something on native).

chrysn commented:
I've added a simple device driver based on an existing driver crate -- for the LSM303AGR, nothing fancy and nothing complete, but nice enough to use with SAUL. It needs a few polishing steps, but already shows open questions for which I could need input from the wider RIOT community:

* How do we want to configure Rust drivers?

  Right now, for an initial demo, I went all the C way, and placed a `boards/microbit-v2/include/lsm303agr-config.rs` with the puny content `const I2C_DEVICES: &[u8] = &[0];` (we have one I2C bus that contains an LSM303AGR, that is bus 0).

  That's not very pretty, but should be familiar in pattern. It's especially not pretty as it looks like it's taken from the include path, but really it's `${BOARDDIR}/include/` because there's no explicit include path known to Rust, although we could make it known for that purpose.

  There's a wide range of alternatives, from just putting the `[0]` in there all the way to parsing config out of some board description file at build time.

* Where do module descriptions and makeadata go to?

  When I started this, I placed all the pseudomodules' dependencies in the catch-all modules Makefile.dep; in hindsight that's a mistake (generally, pulling an external module as first example was) and each module should do it on its own. (Can pseudomodules have Makefile.dep etc?).

  Also, how do we document Rust modules? So far, I've written a doc.txt. I'm not sure we need anything more fancy (given that the interface is what needs to be documented, and the implementation should be self-explanatory or can rely on comments).

* How do place authorship and copyright information on Rust modules?

  The current RIOT practice of having doxygen comments and license headers works badly for Rust (doxygen not being parsed, and the headers being uncommon). Instead, the typical Rust way is to place them in machine readable form in Cargo.toml.

  I recommend we do the latter, and if we do want license information per file, a SPDX entry would do.

chrysn commented:
Recent commits just enhance the quality of the produced driver (but depend on unreleased versions of riot-wrappers and are thus not build-tested again); general status is still that the posed questions need answering for the concrete module, and that I'll need to add Kconfig modelling to ensure that the equivalency tests pass.

chrysn commented:
Updated once more, now with at least one test listed in the PR description.

For better testing (and to get rid of the ugly do-not-commit commit), it'd be helpful to get https://github.com/RIOT-OS/RIOT/pull/17554 in first -- as that'll create a situation where a Rust application will also use a Rust library (and has to cater for that, as the alternative is linking Rust twice and that's a recipe for inefficient use of program memory).

chrysn commented:
I have a plan now (which I can't put in action right away due to time constraints) on how to solve the remaining two issues:

* All RIOT modules written in Rust need to be crates with that module's name (or possibly with some prefix).
* Something in rust_lib_catchall goes through all potential module directories and builds Cargo.toml with a `$module = { path = "$path_to_module", optional = true }` depends line, and something to be included into lib.rs like `#[cfg(feature = "$module")]` / `pub use $module;` line. (I think the `pub use` will have the same effect as the current `extern crate`; if not then there also needs to be an extern crate to keep XFA usable).

  That something will need mechanism to avoid pulling itself in.

* Optionally, if features of crates should become pseudomodules, the Makefile would add `$pseudomodulename = [ "$mainmodule/$pseudomodulename" ]` lines as well. (Could even just parse the crate's `Makefile.include` for `PSEUDOMODULES += ...` lines).

This provides that
* modules are discovered automatically (including in external module directories)
* (by going through all possible and not only loaded module dirs) the Cargo.lock won't wobble around when modules are enabled or disabled
* code is usable by other Rust code: that code would depend on `riot_lib_catchall/$module` (ok, `riot_lib_catchall` needs a rename), and use the Rust API through the `pub use`. (C code could, as it can now, use exported C functions through extern functions with locally copied signatures, or through any included headers the crate might provide. As the original Rust API is available, even if it's just used in the form of the `extern "C"` functions, riot-sys will not need to build wrappers for that).

chrysn commented:
This is now updated to what I consider a much more usable state:

* Some of the plan in https://github.com/RIOT-OS/RIOT/pull/16833#issuecomment-1071243210 is enacted -- missing is automation in enumerating, but as long as it's only 2 modules I don't think that has high priority as long as all the information is somewhere in the tree for building such an automation (which is the case now).
* I've split (and suitably named) the two functionalities into crates/modules:
  * Collecting any Rust modules is the task of `rust_riotmodules` crate. (The name was chosen because it's about RIOT modules written in Rust, and it emphasizes "riot" in the name because in the Rust context modules are something different). This is where all the finding crates, enabling them depending on enabled Rust modules etc. is happening. This is used both by Rust and C applications.
  * Providing a Rust "runtime" (really, just a root crate with a simple panic handler) happens in `rust_riotmodules_standalone`. When applications do *not* have an own root crate but depend on Rust modules, then this gets enabled and provides the entry point to the Rust build infrastructure.
* I've done away with the old external riot-shell-commands example library -- using an external library without a place in the RIOT tree would be an odd corner case. To still have something easily testable, I've implemented a "hello world" command in a Rust module. It is placed in shell/democommands purely for easy separation from the regular shell/commands -- any Rust commands could easily live there, but the demo "hello world" does not have an actual module to depend on, and there'd be many more Makefile lines just to introduce a feature that then guides which parts of the module would be built. (But if reviewers prefer I can just integrate it there, it's to some extent probably also a matter of preference).

There is one prominent "DO NOT MERGE" commit in there that enables SAUL in the rust-gcoap example, just to verify that Rust applications also work well when a Rust module is depended on. I'd very open to any ideas as to how to make this less forced (but that might need a third Rust demo application, maybe SAUL related because that's what the meaningful Rust module does).

I think this is ready for a round of reviews.

chrysn commented:
If reviewers prefer to review the lsm303agr sensor driver independently, I'll be happy to split that out -- but I think that it's valuable to have it in here at least in the review phase as it well illustrates [in a single commit](https://github.com/RIOT-OS/RIOT/pull/16833/commits/471dd4cc0fbc99e840cc771064b9632c651f68f8) what it takes to add a driver in Rust.

chrysn commented:
Addressing recent build failues:
* The rust-gcoap example grew (in a DO NOT MERGE commit, see above) -- a fixup for that now puts them on the "too tight" list.
* A few esp32-wroom failures are probably unrelated (that happens ever now and then).
* "Hashes of binaries with and without Kconfig mismatch for tests/saul with microbit-v2" -- I could need some help modelling this all in Kconfig.
* tests/thread_float on samr21-xpro timed out, probably just a glitch.

MrKevinWeiss reviewed (commented):
From the Kconfig perspective, I don't think we need to expose the `APPLICATION_RUST_MODULE` if we just manually turn off the `MODULE_RUST_RIOTMODULES_STANDALONE` in the app.test.config.

If the one-liner is too much, we could think about adding a flag to the `kconfig.mk` similar to the `RIOT_CONFIG_DEVELHELP`, however, I don't know how scalable this solution would be as we are injecting some make dependency resolution aspects into kconfig.

<details><summary>diff</summary>

```diff
diff --git a/boards/microbit-v2/Kconfig b/boards/microbit-v2/Kconfig
index 5b6d68fe28..44eacaf949 100644
--- a/boards/microbit-v2/Kconfig
+++ b/boards/microbit-v2/Kconfig
@@ -19,6 +19,7 @@ config BOARD_MICROBIT_V2
     select HAS_VDD_LC_FILTER_REG1
 
     select HAVE_SAUL_GPIO
+    select HAVE_LSM303AGR
 
 source "$(RIOTBOARD)/common/microbit/Kconfig"
 source "$(RIOTBOARD)/common/nrf52/Kconfig"
diff --git a/drivers/Kconfig b/drivers/Kconfig
index e93c467c7a..3efeedec0f 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -106,6 +106,7 @@ rsource "lpd8808/Kconfig"
 rsource "lpsxxx/Kconfig"
 rsource "lsm6dsl/Kconfig"
 rsource "lsm303dlhc/Kconfig"
+rsource "lsm303agr/Kconfig"
 rsource "ltc4150/Kconfig"
 rsource "mag3110/Kconfig"
 rsource "mhz19/Kconfig"
diff --git a/drivers/lsm303agr/Kconfig b/drivers/lsm303agr/Kconfig
new file mode 100644
index 0000000000..089158ac31
--- /dev/null
+++ b/drivers/lsm303agr/Kconfig
@@ -0,0 +1,20 @@
+# Copyright (c) 2022 HAW Hamburg
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+
+config MODULE_LSM303AGR
+    bool
+    prompt "LSM303AGR 3D accelerometer/magnetometer" if !(MODULE_SAUL_DEFAULT && HAVE_LSM303AGR)
+    default y if (MODULE_SAUL_DEFAULT && HAVE_LSM303AGR)
+    depends on HAS_PERIPH_I2C
+    depends on TEST_KCONFIG
+    select  MODULE_RUST_RIOTMODULES
+    select MODULE_PERIPH_I2C
+
+config HAVE_LSM303AGR
+    bool
+    help
+        Indicates that a lsm303agr sensor is present.
diff --git a/examples/rust-hello-world/app.config.test b/examples/rust-hello-world/app.config.test
new file mode 100644
index 0000000000..35e6e62250
--- /dev/null
+++ b/examples/rust-hello-world/app.config.test
@@ -0,0 +1,2 @@
+CONFIG_MODULE_RUST_RIOTMODULES=y
+CONFIG_MODULE_RUST_RIOTMODULES_STANDALONE=n
diff --git a/sys/Kconfig b/sys/Kconfig
index 37943f299b..2b1e75d678 100644
--- a/sys/Kconfig
+++ b/sys/Kconfig
@@ -79,6 +79,7 @@ rsource "progress_bar/Kconfig"
 rsource "ps/Kconfig"
 rsource "random/Kconfig"
 rsource "rtc_utils/Kconfig"
+rsource "rust_riotmodules/Kconfig"
 rsource "saul_reg/Kconfig"
 rsource "schedstatistics/Kconfig"
 rsource "sema/Kconfig"
diff --git a/sys/rust_riotmodules/Kconfig b/sys/rust_riotmodules/Kconfig
new file mode 100644
index 0000000000..a289a26506
--- /dev/null
+++ b/sys/rust_riotmodules/Kconfig
@@ -0,0 +1,19 @@
+# Copyright (c) 2022 HAW Hamburg
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+
+menuconfig MODULE_RUST_RIOTMODULES
+    bool "RUST RIOT modules"
+    depends on TEST_KCONFIG
+    depends on HAS_RUST_TARGET
+    help
+        This module is used when some module asking to have its code built
+        through rust_riotmodules
+
+config MODULE_RUST_RIOTMODULES_STANDALONE
+    bool "RUST RIOT modules are standalone"
+    default y
+    depends on MODULE_RUST_RIOTMODULES
```
</details>

chrysn commented:
Thanks, distributed over fixups to give it a try.

Ad the one-liner, would it help to express this as rust_riotmodules depending on either rust_application (preferred) or rust_riotmodules_standalone, but the former is only provided by the applications that have a main module?

gschorcht commented:
The test run of `tesst/periph_timer` must fail on `esp32-wroom-32` because `TIMER_DEV(0)` is occupied by `ztimer_usec` which is enabled by the module `libc_gettimeofday` which is enabled by default :worried:

maribu commented on drivers/lsm303agr/doc.txt:
```diff
@@ -0,0 +1,47 @@
+Data values are obtained on demand
+whenever queried through @ref drivers_saul.
+
+For each device, a two SAUL entry are registered labelled "LSM303AGR accelerometer" and "LSM303AGR magnetometer",
```
```suggestion
For each device, two SAUL entry are registered labelled "LSM303AGR accelerometer" and "LSM303AGR magnetometer",
```

maribu reviewed (commented):
looks good to me :)

chrysn commented on drivers/lsm303agr/doc.txt:
```diff
@@ -0,0 +1,47 @@
+Data values are obtained on demand
+whenever queried through @ref drivers_saul.
+
+For each device, two SAUL entry are registered labelled "LSM303AGR accelerometer" and "LSM303AGR magnetometer",
```
```suggestion
For each device, two SAUL entries are registered labelled "LSM303AGR accelerometer" and "LSM303AGR magnetometer",
```

kaspar030 commented on drivers/lsm303agr/src/lib.rs:
```diff
@@ -0,0 +1,123 @@
+#![no_std]
```
I think this'll need a license header.

kaspar030 replied:
or maybe not, the rust crates specify their license in Cargo.toml.

chrysn replied:
Yes; that's what also the relevant tools tap into (eg. `cargo).

kaspar030 commented on sys/shell/democommands/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+use riot_wrappers::println;
+use core::fmt::Write;
+
+riot_wrappers::static_command!(static_hello_world, "hello-world", "Print a greeting", hello_world);
```
I think the C macro doesn't allow hyphens, or specifying both a function name and the command name. I don't mind, but maybe, for consistency, rename this to "hello_world"?

chrysn replied:
Sure, adjusted.

The Rust macro tries to be less auto-magic than the C macro (we could stringify, but I chose not to) because that's a kind of automatism I don't think is idiomatic in Rust, but that shouldn't stop from being consistent with the C versions in examples..

chrysn commented:
@kaspar030, are you happy with the adjustment?

Rerunning tests given the failed one looks like a glitch (being unrelated and not having failed in other recent runs)

kaspar030 reviewed (approved):
ACK.

kaspar030 commented:
Please squash!

chrysn commented:
Thanks! Squashed.

In removing the "REMOVE ME" commits, I split one of them (the one that
made the rust-gcoap example use other Rust modules) off and kept half of
it as the now most recent commit: With how the rust_riotmodules
developed since those lines were written, it's now best to just use the
riot_rustmodules crate all the time (the usual no-op has no deployment
impact), as to tolerate any Rust-written RIOT modules that might be
activated as they come in (eg. if saul_default gains support for a
board's sensor).

While murdock is working this off, I'll run another round of self-review
whether everything that's in the commits is still in the righ place, and
still makes sense.

chrysn commented:
Updated again to fix some commit wordings, and documentation typos or misreferences.

chrysn commented:
[Builds and tests passed]. I missed squashing after the latest fixup I mentioned in https://github.com/RIOT-OS/RIOT/pull/16833#issuecomment-1179594512 (https://github.com/RIOT-OS/RIOT/pull/16833/commits/e4279bdc588cc2b36070ab9c2f85d9b776eb7c31). Squashing that in and skipping builds/tests due to the pre- and post-squash having identical tree hashes.

kaspar030 commented:
Nice!
